pennylane/math/ode.py

```diff
@@ -0,0 +1,246 @@
+
+    Solves the initial value problem (IVP) of an ordinary differential equation (ODE)
+
+    .. math:: \frac{dy}{dt} = f(y, t), y(t_0) = y_0
+
+    using an `explicit Runge-Kutta method <https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods>`_.
+    In particular, it is using the 4th order term of the `Dormand Prince <https://en.wikipedia.org/wiki/Dormand%E2%80%93Prince_method>`_ Butcher table.
+
+    Args:
+        func (callable): ``f(y, t, *args)`` defining the ODE. Needs to return a ``ndarray`` of the same shape as ``y``.
+        y0 (ndarray): initial value ``y(t0) = y0``
+        ts (ndarray): finite time steps for the ODE solver to take.
+
+    .. note::
+
+        Unlike many standard implementations, this ``odeint`` solver does not adaptively choose the step
+        sizes, but rather a fix list of times for evaluation have to be provided in ``ts``. For an adaptive step
+        size solver, we refer to `jax.experimental.ode.odeint <https://github.com/google/jax/blob/main/jax/experimental/ode.py>`_.
+
+    .. warning::
+
+        This function is written for ``jax`` only and will not work with other machine learning frameworks typically encountered in PennyLane.
+
+    **Example**
+
+    Let us look at the very simple initial value problem $dy/dt = y,$ $y(0)=3$.
+    The analytic solution is $y(t) = 3e^t$. Say we are interested in the solution at
+    time $t=2$, i.e. $y(2)=3e^2$. We can compute this numerically using ``odeint``
+
+    .. code:: python3
+
+        t = jnp.linspace(0, 2, 20)
+
+        def fun(y, t):
+            return y
+
+        y0 = jnp.array((3.))
+        res = qml.math.odeint(fun, y0, t)
+
+    >>> print(qml.math.isclose(res, 3*jnp.exp(2)))
+    True
+
+    As long as ``fun(y, t)`` outputs arrays of the same shape as ``y``, we can use arbitrary input
+    shapes for ``y0``.
+
+    >>> y0 = jnp.ones((5), dtype=float)
+    >>> y1 = qml.math.odeint(fun, y0, t)
+    >>> print(y0.shape, y1.shape)
+    (5,) (5,)
+
+    We can solve the time-dependent Schrodinger equation for the unitary evolution operator
+
+    .. math:: \frac{d}{dt}U = -i H(t) U
+
+    for a time-dependent Hamiltonian :math:`H(t) = X_0 X_1 + v \sin(t) Z_0 Y_1`
+    in the time window ``(t0, t1) = (0, 4)`` using ``odeint`` in the following way:
+
+    .. code:: python3
+
+        XX = qml.matrix(qml.PauliX(0) @ qml.PauliX(1))
+        ZY = qml.matrix(qml.PauliZ(0) @ qml.PauliY(1))
+
+        ts = jnp.linspace(0., 2., 50)
+        y0 = jnp.eye(2**2, dtype=complex)
+        params = jnp.ones(1, dtype=complex)
+
+        def func(y, t, params):
+            H = jnp.array([XX + params[0] * jnp.sin(t) * ZY])
+            return -1j * jnp.sum(H, axis=0) @ y
+
+        U = qml.math.odeint(func, y0, ts, params)
+
+    Formally, this solution can be written as the time-ordered exponential
+    :math:`U(t_0, t_1) = \text{Texp}\left[-i \int_{t_0}^{t_1}d\tau H(\tau)\right]`
+    (see `Dyson Series <https://en.wikipedia.org/wiki/Dyson_series>`_).
```

Comment:
Yes, honestly I dont know why I put this since I more and more realize what a misleading notation the $\text{Texp}$ is and think it should not be used! 😆 

(happy to comment further offline, but see e.g. here: https://github.com/PennyLaneAI/pennylane/pull/3617#discussion_r1069169255 )